File: dart_sample/lib/demos/general/dismissible_size_changed_layout_demo_page.dart
import 'package:flutter/material.dart';

import '../../counter_widgets.dart';

class DismissibleSizeChangedLayoutDemoPage extends StatefulWidget {
  const DismissibleSizeChangedLayoutDemoPage({super.key});

  @override
  State<DismissibleSizeChangedLayoutDemoPage> createState() =>
      _DismissibleSizeChangedLayoutDemoPageState();
}

class _DismissibleSizeChangedLayoutDemoPageState
    extends State<DismissibleSizeChangedLayoutDemoPage> {
  final List<int> _items = <int>[1, 2, 3];
  int _sizeNotifications = 0;
  bool _expanded = false;
  bool _rightToLeft = false;

  @override
  Widget build(BuildContext context) {
    final TextDirection textDirection = _rightToLeft
        ? TextDirection.rtl
        : TextDirection.ltr;
    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      spacing: 12,
      children: <Widget>[
        const Text(
          'Dismissible + SizeChangedLayoutNotifier',
          style: TextStyle(fontSize: 20, color: Colors.black),
        ),
        const Text(
          'Swipe rows in either direction. The resize probe reports '
          'notifications only after its established layout size changes.',
          style: TextStyle(fontSize: 14, color: Colors.black54),
        ),
        Row(
          spacing: 8,
          children: <Widget>[
            _buildButton(
              label: _rightToLeft ? 'Direction: RTL' : 'Direction: LTR',
              onTap: () {
                setState(() {
                  _rightToLeft = !_rightToLeft;
                });
              },
            ),
            _buildButton(
              label: _expanded ? 'Shrink probe' : 'Grow probe',
              onTap: () {
                setState(() {
                  _expanded = !_expanded;
                });
              },
            ),
            _buildButton(
              label: 'Reset rows',
              onTap: () {
                setState(() {
                  _items
                    ..clear()
                    ..addAll(<int>[1, 2, 3]);
                });
              },
            ),
          ],
        ),
        NotificationListener<SizeChangedLayoutNotification>(
          onNotification: (SizeChangedLayoutNotification notification) {
            WidgetsBinding.instance.addPostFrameCallback((Duration _) {
              if (mounted) {
                setState(() {
                  _sizeNotifications++;
                });
              }
            });
            return false;
          },
          child: Align(
            alignment: Alignment.centerLeft,
            child: SizeChangedLayoutNotifier(
              child: Container(
                width: _expanded ? 320 : 190,
                height: _expanded ? 64 : 44,
                color: const Color(0xFFDCEAF7),
                alignment: Alignment.center,
                child: Text(
                  'layout notifications: $_sizeNotifications',
                  style: const TextStyle(color: Color(0xFF174A72)),
                ),
              ),
            ),
          ),
        ),
        Expanded(
          child: Directionality(
            textDirection: textDirection,
            child: ListView(
              children: _items.map(_buildDismissibleRow).toList(),
            ),
          ),
        ),
      ],
    );
  }

  Widget _buildDismissibleRow(int item) {
    return Padding(
      padding: const EdgeInsets.only(bottom: 8),
      child: Dismissible(
        key: ValueKey<int>(item),
        background: _buildRowBackground(
          'START →',
          Alignment.centerLeft,
          const Color(0xFF2E7D32),
        ),
        secondaryBackground: _buildRowBackground(
          '← END',
          Alignment.centerRight,
          const Color(0xFFC62828),
        ),
        crossAxisEndOffset: 0.08,
        onDismissed: (DismissDirection direction) {
          setState(() {
            _items.remove(item);
          });
        },
        child: _buildRowSurface('Swipe row $item', const Color(0xFFF4F6F8)),
      ),
    );
  }

  static Widget _buildRowSurface(String label, Color color) {
    return Container(
      height: 58,
      color: color,
      padding: const EdgeInsets.symmetric(horizontal: 16),
      alignment: Alignment.centerLeft,
      child: Text(label, style: const TextStyle(color: Colors.black)),
    );
  }

  static Widget _buildRowBackground(
    String label,
    Alignment alignment,
    Color color,
  ) {
    return Container(
      height: 58,
      color: color,
      padding: const EdgeInsets.symmetric(horizontal: 16),
      alignment: alignment,
      child: Text(label, style: const TextStyle(color: Colors.white)),
    );
  }

  static Widget _buildButton({
    required String label,
    required VoidCallback onTap,
  }) {
    return SizedBox(
      width: 130,
      child: CounterTapButton(
        label: label,
        onTap: onTap,
        background: const Color(0xFFDCE3ED),
        foreground: Colors.black,
        fontSize: 12,
        padding: const EdgeInsets.symmetric(horizontal: 8, vertical: 7),
      ),
    );
  }
}

File: src/Sample/Plumix.Sample/Demos/General/DismissibleSizeChangedLayoutDemoPage.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Avalonia;
using Avalonia.Media;
using Plumix.Foundation;
using Plumix.Rendering;
using Plumix.UI;
using Plumix.Widgets;

namespace Plumix;

// Dart parity source: dart_sample/lib/demos/general/dismissible_size_changed_layout_demo_page.dart

public sealed class DismissibleSizeChangedLayoutDemoPage : StatefulWidget
{
    public override State CreateState() => new DismissibleSizeChangedLayoutDemoPageState();
}

internal sealed class DismissibleSizeChangedLayoutDemoPageState : State
{
    private readonly List<int> _items = [1, 2, 3];
    private int _sizeNotifications;
    private bool _expanded;
    private bool _rightToLeft;

    public override Widget Build(BuildContext context)
    {
        TextDirection textDirection = _rightToLeft ? TextDirection.Rtl : TextDirection.Ltr;
        return new Column(
            crossAxisAlignment: CrossAxisAlignment.Stretch,
            spacing: 12,
            children:
            [
                new Text("Dismissible + SizeChangedLayoutNotifier", fontSize: 20, color: Colors.Black),
                new Text(
                    "Swipe rows in either direction. The resize probe reports notifications only after its " +
                    "established layout size changes.",
                    fontSize: 14,
                    color: Colors.DimGray),
                new Row(
                    spacing: 8,
                    children:
                    [
                        BuildButton(
                            _rightToLeft ? "Direction: RTL" : "Direction: LTR",
                            () => SetState(() => _rightToLeft = !_rightToLeft)),
                        BuildButton(
                            _expanded ? "Shrink probe" : "Grow probe",
                            () => SetState(() => _expanded = !_expanded)),
                        BuildButton(
                            "Reset rows",
                            () => SetState(() =>
                            {
                                _items.Clear();
                                _items.AddRange([1, 2, 3]);
                            })),
                    ]),
                new NotificationListener<SizeChangedLayoutNotification>(
                    onNotification: _ =>
                    {
                        Scheduler.AddPostFrameCallback(_ =>
                        {
                            if (Mounted)
                            {
                                SetState(() => _sizeNotifications++);
                            }
                        });
                        return false;
                    },
                    child: new Align(
                        alignment: Alignment.CenterLeft,
                        child: new SizeChangedLayoutNotifier(
                            child: new Container(
                                width: _expanded ? 320 : 190,
                                height: _expanded ? 64 : 44,
                                color: Color.Parse("#FFDCEAF7"),
                                alignment: Alignment.Center,
                                child: new Text(
                                    $"layout notifications: {_sizeNotifications}",
                                    color: Color.Parse("#FF174A72")))))),
                new Expanded(
                    child: new Directionality(
                        textDirection,
                        new ListView(
                            children: _items
                                .Select(BuildDismissibleRow)
                                .ToArray()))),
            ]);
    }

    private Widget BuildDismissibleRow(int item)
    {
        return new Padding(
            insets: new Thickness(0, 0, 0, 8),
            child: new Dismissible(
                key: new ValueKey<int>(item),
                child: BuildRowSurface($"Swipe row {item}", Color.Parse("#FFF4F6F8")),
                background: BuildRowBackground("START →", Alignment.CenterLeft, Color.Parse("#FF2E7D32")),
                secondaryBackground: BuildRowBackground(
                    "← END",
                    Alignment.CenterRight,
                    Color.Parse("#FFC62828")),
                crossAxisEndOffset: 0.08,
                onDismissed: _ => SetState(() => _items.Remove(item))));
    }

    private static Widget BuildRowSurface(string label, Color color)
    {
        return new Container(
            height: 58,
            color: color,
            padding: new Thickness(16, 0),
            alignment: Alignment.CenterLeft,
            child: new Text(label, color: Colors.Black));
    }

    private static Widget BuildRowBackground(string label, Alignment alignment, Color color)
    {
        return new Container(
            height: 58,
            color: color,
            padding: new Thickness(16, 0),
            alignment: alignment,
            child: new Text(label, color: Colors.White));
    }

    private static Widget BuildButton(string label, Action onTap)
    {
        return new SizedBox(
            width: 130,
            child: new CounterTapButton(
                label: label,
                onTap: onTap,
                background: Color.Parse("#FFDCE3ED"),
                foreground: Colors.Black,
                fontSize: 12,
                padding: new Thickness(8, 7)));
    }
}
